

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>File Formats &mdash; Megalodon 2.0.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Variant Phasing" href="variant_phasing.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Megalodon
          

          
          </a>

          
            
            
              <div class="version">
                2.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="algorithm_details.html">Megalodon Algorithm Details</a></li>
<li class="toctree-l1"><a class="reference internal" href="common_arguments.html">Common Arguments</a></li>
<li class="toctree-l1"><a class="reference internal" href="advanced_arguments.html">Advanced Megalodon Arguments</a></li>
<li class="toctree-l1"><a class="reference internal" href="computing_considerations.html">Computing Considerations</a></li>
<li class="toctree-l1"><a class="reference internal" href="variant_phasing.html">Variant Phasing</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">File Formats</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#base-calling">Base Calling</a></li>
<li class="toctree-l2"><a class="reference internal" href="#mapping">Mapping</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#whatshap-mapping">Whatshap Mapping</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#per-read-modified-bases">Per-read Modified Bases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#database">Database</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tab-delimited">Tab-delimited</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#aggregated-modified-bases">Aggregated Modified Bases</a></li>
<li class="toctree-l2"><a class="reference internal" href="#per-read-sequence-variants">Per-read Sequence Variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="#aggregated-sequence-variants">Aggregated Sequence Variants</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Megalodon</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>File Formats</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/file_formats.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="file-formats">
<h1>File Formats<a class="headerlink" href="#file-formats" title="Permalink to this headline">¶</a></h1>
<p>This page describes the output file formats produced by <code class="docutils literal notranslate"><span class="pre">megalodon</span></code>.</p>
<div class="section" id="base-calling">
<h2>Base Calling<a class="headerlink" href="#base-calling" title="Permalink to this headline">¶</a></h2>
<p>Basecalling produces either FASTQ or FASTA formats.
Basecalls will be output into the <code class="docutils literal notranslate"><span class="pre">basecalls.fastq</span></code> or <code class="docutils literal notranslate"><span class="pre">basecalls.fasta</span></code> file within the <code class="docutils literal notranslate"><span class="pre">--output-directory</span></code>.</p>
<p>Basecall anchored modified base calls are output into a custom HDF5 format (similar to the guppy output format descibed on the community page).
The HDF5 format contains a single dataset, <code class="docutils literal notranslate"><span class="pre">mod_long_names</span></code> at the root level which contains the modified base long names as described in the model used for calling.
The <code class="docutils literal notranslate"><span class="pre">Reads</span></code> group contains all of the per-read modified base scores.
Within this group each reads modified base scores are stored in a dataset indexed by the read_id.
This dataset contains an array with dimensions <code class="docutils literal notranslate"><span class="pre">basecall</span> <span class="pre">length</span></code> by <code class="docutils literal notranslate"><span class="pre">number</span> <span class="pre">of</span> <span class="pre">modified</span> <span class="pre">bases</span></code>.
Unlike the guppy output, scores are only recorded where applicable given the basecall made (e.g. 5mC calls are only output at canonical C basecall positions).
Invalid positions are represented with a <code class="docutils literal notranslate"><span class="pre">NAN</span></code> value.
Note that the modified base scores outside of the dependent canonical contexts are not effected during training, so these values should not be used from the guppy output.</p>
</div>
<div class="section" id="mapping">
<h2>Mapping<a class="headerlink" href="#mapping" title="Permalink to this headline">¶</a></h2>
<p>Mapped reads can be output in SAM, BAM or CRAM formats.
Basecalls will be output into the <code class="docutils literal notranslate"><span class="pre">mappings.sam</span></code>, <code class="docutils literal notranslate"><span class="pre">mappings.bam</span></code>, or <code class="docutils literal notranslate"><span class="pre">mappings.cram</span></code> file within the <code class="docutils literal notranslate"><span class="pre">--output-directory</span></code>.</p>
<div class="section" id="whatshap-mapping">
<h3>Whatshap Mapping<a class="headerlink" href="#whatshap-mapping" title="Permalink to this headline">¶</a></h3>
<p>In addition to standard mapping files, megalodon includes a special mapping-style output with specific relevence to the variant calling pipeline.
This format can be output as a SAM, BAM or CRAM file as with standard mapping format.
The mapped reads in this output represent only the information about proposed variants contained within each read.
Each read is first replaced with the reference bases, with only positions with proposed variants replaced with variants calls.
The score for each call is encoded in the base quality scores for each read.
This output is useful for 1) producing more accuracte variant phasing and read haplotagging via whatshap and 2) visualizing per-read variant calls in a genome browser.</p>
</div>
</div>
<div class="section" id="per-read-modified-bases">
<h2>Per-read Modified Bases<a class="headerlink" href="#per-read-modified-bases" title="Permalink to this headline">¶</a></h2>
<div class="section" id="database">
<h3>Database<a class="headerlink" href="#database" title="Permalink to this headline">¶</a></h3>
<p>The primary output for per-read modified base results is an <a class="reference external" href="https://www.sqlite.org/index.html">sqlite database</a>.
This database contains an indexed table with per-read, per-position, modified base scores, as well as auxiliary tables with read, modification type and reference position information.
The read table (<code class="docutils literal notranslate"><span class="pre">read</span></code>) contains the read UUID.
The modification type table (<code class="docutils literal notranslate"><span class="pre">mod</span></code>) contains the single letter modified base code, the location sequence match motif, the raw (including ambiguous bases) motif, and the relative modified base position within the motifs.
The reference position table (<code class="docutils literal notranslate"><span class="pre">pos</span></code>) contains the mapped 0-based position, strand (1=forward, -1=reverse) and chromosome (via a final <code class="docutils literal notranslate"><span class="pre">chrm</span></code> table which contains the chromosome text).</p>
<p>This database may be accessed via the <code class="docutils literal notranslate"><span class="pre">megalodon.mods.ModsDb</span></code> object.</p>
</div>
<div class="section" id="tab-delimited">
<h3>Tab-delimited<a class="headerlink" href="#tab-delimited" title="Permalink to this headline">¶</a></h3>
<p>Modified bases results are also available via tab-dellimited text output.
This output can be requested via the <code class="docutils literal notranslate"><span class="pre">--write-mods-text</span></code> flag.
This output contains the following fields: <code class="docutils literal notranslate"><span class="pre">read_id</span></code>, <code class="docutils literal notranslate"><span class="pre">chrm</span></code>, <code class="docutils literal notranslate"><span class="pre">strand</span></code>, <code class="docutils literal notranslate"><span class="pre">pos</span></code>, <code class="docutils literal notranslate"><span class="pre">mod_log_prob</span></code>, <code class="docutils literal notranslate"><span class="pre">can_log_prob</span></code>, <code class="docutils literal notranslate"><span class="pre">mod_base</span></code>, <code class="docutils literal notranslate"><span class="pre">motif</span></code></p>
</div>
</div>
<div class="section" id="aggregated-modified-bases">
<h2>Aggregated Modified Bases<a class="headerlink" href="#aggregated-modified-bases" title="Permalink to this headline">¶</a></h2>
<p>The default aggregated modified base output is the bedMethyl format (<a class="reference external" href="https://www.encodeproject.org/data-standards/wgbs/">description here</a>).
Alternative formats are <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/wiggle.html">wiggle</a> (variableStep) and VCF (treating the modified base as if it were a sequence variant).</p>
</div>
<div class="section" id="per-read-sequence-variants">
<h2>Per-read Sequence Variants<a class="headerlink" href="#per-read-sequence-variants" title="Permalink to this headline">¶</a></h2>
<p>As with the modified base results, the primary output for per-read sequence variant results is as <a class="reference external" href="https://www.sqlite.org/index.html">sqlite database</a>.
This database contains an indexed table with per-read, per-position, variant scores, as well as auxiliary tables with read, reference location and alternative allele information.</p>
<p>The reference location table (<code class="docutils literal notranslate"><span class="pre">loc</span></code>) contains the mapped 0-based position, strand (1=forward, -1=reverse) and chromosome (via a final <code class="docutils literal notranslate"><span class="pre">chrm</span></code> table which contains the chromosome text).
The <code class="docutils literal notranslate"><span class="pre">loc</span></code> table also contains the location for the start and end of the tested positions (applicable for insertions/delections).
For example, insertions generally require a context base for downstream processing, but within megalodon only the inserted position is considered (without context).
Each reference location is linked to the IDs linked with this location from the input variants file.
Finally the reference sequence for the location is included in this table.
In the related <code class="docutils literal notranslate"><span class="pre">alt</span></code> table, each alternative sequence is stored.
Links between alternative sequences and reference locations are made via the main <code class="docutils literal notranslate"><span class="pre">data</span></code> table.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">read</span></code> table contrains the read UUID as well as the mapped strand for each read.</p>
</div>
<div class="section" id="aggregated-sequence-variants">
<h2>Aggregated Sequence Variants<a class="headerlink" href="#aggregated-sequence-variants" title="Permalink to this headline">¶</a></h2>
<p>Sequence variant calls are output in standard VCF format (version 4.1).
The sample format fields includes the following standard VCF fields: <code class="docutils literal notranslate"><span class="pre">gt</span></code>, <code class="docutils literal notranslate"><span class="pre">gq</span></code>, <code class="docutils literal notranslate"><span class="pre">gp</span></code>, <code class="docutils literal notranslate"><span class="pre">gl</span></code>, and <code class="docutils literal notranslate"><span class="pre">``pl</span></code>
In addition the non-standard <code class="docutils literal notranslate"><span class="pre">log_probs</span></code> field, containing the per-read contributions to the variant call, can be added to the VCF file by setting the <code class="docutils literal notranslate"><span class="pre">--write-vcf-log-probs</span></code> flag.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="variant_phasing.html" class="btn btn-neutral float-left" title="Variant Phasing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Oxford Nanopore Technologies

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>